contracts/lqmining/parallel/LMPoolSelfHosted.sc

```diff
@@ -0,0 +1,288 @@
+        (poolVLQ0._1, releasedVLQ) == bundleOut.tokens(0) &&
+        (poolTMP0._1, releasedTMP) == bundleOut.tokens(1) &&
+        bundleOut.R6[Coll[Byte]].get == poolNFT0._1
+
+      prevEpochsCompounded &&
+      epochsAllocated >= 2 &&
+      lastCompoundedEpochPreserved &&
+      deltaLQ == -deltaVLQ &&
+      releasedTMP == -deltaTMP &&
+      validBundle
+
+    } else if (deltaLQ < 0) { // Redeem.
+      val releasedLQ = deltaVLQ
+      val minReturnedTMP =
+        if (curEpochIx > epochNum) 0L
+        else {
+          val epochsDeallocated = epochNum - max(0L, curEpochIx)
+          releasedLQ * epochsDeallocated
+        }
+
+      val redeemNoLimit = HEIGHT >= programStart + epochNum * epochLen + redeemLimitDelta
+
+      (prevEpochsCompounded || redeemNoLimit) &&
+      lastCompoundedEpochPreserved &&
+      (deltaVLQ == -deltaLQ) &&
+      (deltaTMP >= minReturnedTMP)
+
+    } else {
+
+      val deltaMainReward = poolMainReward1._2 - reservesMainReward
+      val deltaOptReward  = poolOptReward1._2 - reservesOptReward
+
+      if (deltaTMP > 0) // Compound.
+        {
+          val epoch            = lastCompoundedEpoch0 + 1
+          val epochsToCompound = epochNum - epoch
+          val epochMainAlloc   = prevMainProgramBudget0 / (epochsToCompound + 1)
+          val epochOptAlloc    = prevOptProgramBudget0 / (epochsToCompound + 1)
+
+          // Epoch is legal:
+          val legalEpoch = epoch <= curEpochIx - 1
+
+          val actualTMP    = 0x7fffffffffffffffL - poolTMP0._2 - reservesLQ * epochsToCompound
+          val allocMainRem = reservesMainReward - epochMainAlloc * epochsToCompound
+          val allocOptRem  = reservesOptReward - epochOptAlloc * epochsToCompound
+
+          if (actualTMP > 0) {
+            val rewardMain = (allocMainRem.toBigInt - 1L) * deltaTMP / actualTMP
+            val rewardOpt  = (allocOptRem.toBigInt - 1L) * deltaTMP / actualTMP
+
+            legalEpoch &&
+            lastCompoundedEpochPreserved &&
+            (-deltaMainReward <= rewardMain) &&
+            (-deltaOptReward <= rewardOpt) &&
+            (deltaLQ == 0L) &&
+            (deltaVLQ == 0L)
+
+          } else false
+        } else if (deltaTMP == 0) {
+
+        val budgets1 = successor.R5[Coll[Long]].get
+
+        val prevMainProgramBudget1 = budgets1(0)
+        val prevOptProgramBudget1  = budgets1(1)
+
+        // Mark if only main reward is available:
+        val optBudgetIsNotInit = prevOptProgramBudget0 == 1L
+
+        // Calculate if pool stores correct budgets:
+        val budgetsNotUpdated =
+          (reservesMainReward - prevMainProgramBudget0 <= -maxRoundingError0) ||
+          (!optBudgetIsNotInit && (reservesOptReward - prevOptProgramBudget0 <= -maxRoundingError0))
+
+        val budgetsNotDepleted = poolMainReward1._2 != 0 && poolOptReward0._2 != 0
+
+        if (
+          (deltaMainReward > 0 || deltaOptReward > 0) && !budgetsNotUpdated && budgetsNotDepleted
```

Comment:
Budget can be increased only in the case opt token is initialized and not depleted?

Reply:
Same, depleted means 0 tokens, which is impossible during normal program operation